src/modules/nathelper/nathelper.c

```diff
@@ -2445,6 +2452,175 @@ static int sel_rewrite_contact(str *res, select_t *s, struct sip_msg *msg)
+	       LM_ERR("Parsing Contact Header is failed\n");
+	       return -1;
+	}
+
+
+
+	contact.s = ((contact_body_t *)msg->contact->parsed)->contacts->name.s;
+	contact.len = ((contact_body_t *)msg->contact->parsed)->contacts->len;
+
+	alias_to_uri(&contact,&alias_uri);
+	write_to_avp(msg, &alias_uri, uri_avp);
+
+	return 1;
+
+}
+/**
+* write_to_avp function writes data to given avp
+* @param data  source data
+* @param uri_avp destination avp name
+*/
+void write_to_avp(struct sip_msg *msg, str *data, str *uri_avp){
+	pv_spec_t *pvresult = NULL;
+	pv_value_t valx;
+	pvresult = pv_cache_get(uri_avp);
+
+	if(pvresult == NULL) {
+	       LM_ERR("Failed to malloc destination pseudo-variable \n");
+	       return;
+	}
+
+	if(pvresult->setf==NULL) {
+	       LM_ERR("destination pseudo-variable is not writable: %.*s \n",uri_avp->len, uri_avp->s);
+	       return;
+	}
+	memset(&valx, 0, sizeof(pv_value_t));
+
+	if(!data->s){
+				 LM_ERR("There isnt any data to write avp\n");
+				 return;
+	}
+
+	valx.flags      =       PV_VAL_STR;
+	valx.rs.s       =       data->s;
+	valx.rs.len     =       data->len;
+
+	LM_DBG("result: %.*s\n", valx.rs.len, valx.rs.s);
+	pvresult->setf(msg, &pvresult->pvp, (int)EQ_T, &valx);
+
+}
+/**
+* function alias_to_uri  select alias paramter from @param contact_header
+* then writes to @param alias_uri
+* @param contact_header  Source contact header
+* @param alias_uri destination string
+*/
+void alias_to_uri(str *contact_header, str *alias_uri){
+	int i=0; // index
+	str host={0,0};
+	str port={0,0};
+	str proto={0,0};
+	char *memchr_pointer;
+
+	if(!contact_header)
+	       return;
+
+	LM_DBG("Contact header [%.*s] \r\n",contact_header->len,contact_header->s);
+
+	for(i=0; i<contact_header->len  ;i++){
+		if(strncmp(&contact_header->s[i], SALIAS, SALIAS_LEN) == 0){
+			i=i+SALIAS_LEN;
+			host.s = &contact_header->s[i];
+			memchr_pointer = memchr(host.s , 126 /* ~ */,contact_header->len-i);
+				if(memchr_pointer == NULL) {
+					LM_ERR("No alias param found for host\n");
+					return;
+				} else {
+					host.len = memchr_pointer - &contact_header->s[i];
+					i=i+host.len;
+				}
+			break;
+			}
+	}
+
+	if(!memchr_pointer){
+		LM_ERR("No alias param found\n");
+		return ;
+	}
+	if(&memchr_pointer[1])
+		port.s=&memchr_pointer[1];
+	else{
+		LM_ERR("We cannot find alias sign for port - exit\n");
+		return;
+	}
+
+	memchr_pointer = memchr(port.s , 126 /* ~ */,contact_header->len-i);
+	if(memchr_pointer == NULL) {
+		LM_ERR("No alias param found for port\n");
+		return ;
+	} else {
+		port.len = memchr_pointer - port.s;
+		i=i+port.len;
+	}
+
+	//last char is proto 0,1,2,3,4..7
+	proto.s= &port.s[port.len+1];
+	proto_type_int_to_str(atoi(proto.s), &proto);
+
+	LM_DBG("Host [%.*s][port: %.*s][proto: %.*s] \r\n",host.len,host.s,port.len,port.s,proto.len,proto.s);
+
+	//sip:host:port;transport=udp
+	alias_uri->s =(char *) pkg_malloc(port.len+host.len+proto.len+16);
```

Comment:
When is alias_uri->s freed?

Reply:
i looked some examples in codes for pv->setf and i dont think that it should be freed.